Add a blog post about Wasm based extensibility progress. (#8916)
istio/istio.io: content/en/blog/2021/wasm-progress/index.md

---
title: Progress of Envoy and Istio WebAssembly Extensibility
description: An update on Envoy and Istio WebAssembly based extensibility effort.
publishdate: 2021-02-09
attribution: "Pengyuan Bian (Google)"
keywords: [wasm,extensibility,WebAssembly]
---

Back to Istio 1.5 release, Istio introduced [WebAssembly based extensibility](/blog/2020/wasm-announce/) support.
Over the past year, Istio, Envoy, and Proxy-Wasm communities have continued the effort to make WebAssembly extensibility stable, reliable, and easy to adopt.
In this blog post, we will walk through the updates in this area at Istio 1.9 release.

## Proxy-Wasm (WebAssembly) support merged in upstream Envoy

The initial development work of Envoy WebAssembly runtime happened at `envoy-wasm` fork.

> Over time, we collected great feedback from community early adopters, as well as Istio first class Wasm extensions development, which helped us to continuously improve the runtime.

Reviewer: ```suggestion
After our initial experimental support was available, we collected great feedback from community early adopters. Along with experience gained from development efforts on core Istio Wasm extensions, this feedback helped us mature and stabilize the runtime.
```

> Since Istio 1.8, WebAssembly support was merged into the Envoy main tree. This is a significant milestone for Istio and Envoy Wasm efforts, since it indicates that:

Reviewer: ```suggestion
These improvements unblocked merging Wasm support directly into Envoy, allowing it to become part of all official Envoy releases (as of the Istio 1.8 release). This is a significant milestone for Istio and Envoy Wasm efforts, since it indicates that:
```

* The runtime is ready for wider adoption.
* The programming ABI/API, extension configuration API, as well as runtime behavior are becoming stable.
* Better community support moving forward.

## Istio Wasm Extension Ecosystem

> As one of the early adopters of Envoy WebAssembly runtime, Istio extensibility working group gained a lot of experience via developing several first class extensions, such as metadata exchange, Prometheus stats, and attribute generation.

Reviewer: Should we link to the WG page on community repo?  And should we link to any docs pages on the extensions (for audiences that might not know what metadata exchange is, for instance) ?

Author: Where is the working group page? I cannot find it in community repo.. and closest I can find is https://github.com/istio/community/blob/master/WORKING-GROUPS.md

> In order to share the learning broadly, a [`wasm-extensions` repository](https://github.com/istio-ecosystem/wasm-extensions) is created under `istio-ecosystem` org. This repository serves two purposes:

Reviewer: ```suggestion
In order to share the learning broadly, a [`wasm-extensions` repository](https://github.com/istio-ecosystem/wasm-extensions) was created under `istio-ecosystem` org. This repository serves two purposes:
```

> As one of the early adopters of Envoy WebAssembly runtime, Istio extensibility working group gained a lot of experience via developing several first class extensions, such as [metadata exchange](/docs/reference/config/proxy_extensions/metadata_exchange/), [Prometheus stats](/docs/reference/config/proxy_extensions/stats/), and [attribute generation](/docs/reference/config/proxy_extensions/attributegen/).

Reviewer: nit: do you want to say `Istio Extensions and Telemetry working group` here for consistency with the group names?

Author: Done.

> * Provides canonical example extensions, which also covers several highly demanded features such as basic authentication.

Reviewer: ```suggestion
* Provides canonical example extensions, which also cover several highly demanded features such as basic authentication.
```

> * Provides canonical example extensions, which also covers several highly demanded features such as basic authentication.

Reviewer: can we link to the basic auth extension?

> * Provides guides about Wasm extension development, testing, and release. The guide is based on the same build tool chains and test frameworks that are used, maintained and tested by the Istio extensibility team.

Reviewer: ```suggestion
* Provides a guide for Wasm extension development, testing, and release. The guide is based on the same build tool chains and test frameworks that are used, maintained and tested by the Istio extensibility team.
```

Currently the existing guide covers [WebAssembly extension development](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-a-wasm-extension-with-cpp.md)
and [unit testing](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-cpp-unit-test.md) with C++,
as well as [integration testing](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-integration-test.md) with a GoLang test framework,
which simulates real runtime by running a Wasm module with the Istio proxy binary.
In the future, several more canonical extensions will also be added such as open policy agent, and header manipulation based on JWT token.

## Wasm Module Distribution via Istio Agent

> Prior to Istio 1.9, to distribute remote Wasm modules to the proxy, [Envoy remote data source](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-remotedatasource) needs to be configured.

Reviewer: ```suggestion
Prior to Istio 1.9, [Envoy remote data source](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-remotedatasource) were needed to distribute remote Wasm modules to the proxy.
```

> Prior to Istio 1.9, [Envoy remote data source](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-remotedatasource) were needed to distribute remote Wasm modules to the proxy.

Reviewer: ```suggestion
Prior to Istio 1.9, [Envoy remote data sources](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-remotedatasource) were needed to distribute remote Wasm modules to the proxy.
```

where two `EnvoyFilter` resources are defined: one is to add a remote fetch Envoy cluster, and the other one is to inject a Wasm filter into HTTP filter chain.

> However this method has a drawback: when Envoy receives a Wasm extension configuration which has a remote fetch, as long as it passes basic protocol validation, Envoy will send an `ACK` to istiod without waiting for the remote fetch to finish.
> If remote fetch fails either due to bad configuration or transient error, Envoy will be stuck with the bad configuration.

Reviewer: ```suggestion
However this method has a drawback: if remote fetch fails either due to bad configuration or transient error, Envoy will be stuck with the bad configuration.
```

> Furthermore, If the filter is configured as [fail closed](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/wasm/v3/wasm.proto#extensions-wasm-v3-pluginconfig), a bad remote fetch will stop Envoy from serving.

Reviewer: ```suggestion
If a Wasm extension is configured as [fail closed](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/wasm/v3/wasm.proto#extensions-wasm-v3-pluginconfig), a bad remote fetch will stop Envoy from serving.
```

To fix this issue, [a fundamental change](https://github.com/envoyproxy/envoy/issues/9447) is needed to Envoy xDS protocol to make it allow asynchronous xDS response.

> Istio 1.9 provides a reliable distribution mechanism out of the box by leveraging the XDS proxy inside istio-agent and Envoy [Extension Configuration Discovery Server](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/extension) (ECDS).

Reviewer: just above we used `xDS`; here we are using `XDS`. should we just stick to `xDS` ?

Author: Thanks for catching this! Done.

The following diagram shows the flow.
Istio-agent intercepts the extension config resource update from istiod, reads the remote fetch hint from it, downloads the Wasm module, and rewrites the ECDS configuration with the path of the downloaded Wasm module.
If download fails, istio-agent will reject the ECDS update and prevent bad configuration reaching Envoy. For more detail, please go through [this guide](/docs/ops/configuration/extensibility/wasm-module-distribution/).

{{< image width="75%"
    link="./architecture-istio-agent-downloading-wasm-module.svg"
    alt="Remote Wasm module fetch flow"
    caption="Remote Wasm module fetch flow"
    >}}

> ## Istio Wasm SIG and Future Work

Reviewer 2: SIG = Working Group?

Author: In my understanding, Wasm SIG is like a subgroup of extensibility and telemetry WG. It has a very focused topic, which is Wasm based extension. @mandarjog we really need something to link here, so that people understands wasm sig better and join if they like.

Reviewer: can we use https://tinyurl.com/istio-wasm-todo ?

Author: Yeah I guess that is the only available detail link. Added a link to it. Thanks!

* **A first-class extension API**: Currently Wasm extensions needs to be injected via Istio's `EnvoyFilter` API. A first-class extension API will make using Wasm with Istio easier, and we expect this to be introduced in Istio 1.10.
* **Distribution artifacts interoperability**: Built on top of Solo.io’s [WebAssembly OCI image spec effort](https://www.solo.io/blog/announcing-the-webassembly-wasm-oci-image-spec/), a standard Wasm artifacts format will make it easy to build, pull, publish, and execute.
* **Container Storage Interface (CSI) based artifacts distribution**: Using istio-agent to distribute modules is easy for adoption, but may not be efficient as each proxy will keep a copy of the Wasm module. As a more efficient solution, with [Ephemeral CSI](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html), a DaemonSet will be provided which could configure storage for pods. Working similarly to a CNI plugin, a CSI driver would fetch the Wasm module out-of-band from the xDS flow and mount it inside the `rootfs` when the pod starts up.

> If you would like to join us, we meet every other week Tuesdays at 2PM Pacific Time. (5PM ET, 11PM CET)

Reviewer 3: Please update this line to have some sort of "how to join/how to get involved"; and also consider putting your meetings on the Istio working groups calendar.

Author: I added the meeting to istio work group calendar. Also created an announcement at discuss.istio.io and linked it here instead of the Google doc.
